Add documentation to Combinators

andrevidela opened the pull request:
I've added doc strings for most of the functions in `Combinators.idr`.

Could you please review, and give feedback?

I also have a question for the `betweenopt` combinator.

The current implementation is `landopt (roptand open p) close`. But doesn't this allow allow the surrounding parsers to be mismatched?

For example `betweenopt (char "(") (exact "a") (char ")")` would match `(a`, wouldn't it?

gallais commented:
> I also have a question for the `betweenopt` combinator.

You are right, this looks like a bug to me. In agdarsec the definition of `between?`
is [the one you describe](https://github.com/gallais/agdarsec/blob/master/src/Text/Parser/Combinators.agda#L191).

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
+||| parser will fail too. If one succeeds, then the following parsers will
+||| not be attempted and the parser will succeed.
+|||
+||| Unindexed signature: `[Parser a] -> Parser a`
```
Should probably say `List (Parser a)` to avoid haskellisms

andrevidela replied:
Oops, my ~~tongue~~ keystrokes slipped

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,342 @@ andbind p q = MkParser $ \mlen, ts =>
+||| Run two parsers one after the other. If any of the two parsers 
+||| fail this fails. The the result of both parsers is returned as 
+||| a pair.
+|||
```
Repetition: `The the`

andrevidela replied:
Fixed

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,342 @@ andbind p q = MkParser $ \mlen, ts =>
+||| attempt the first parser on the input, and run the function on the
+||| parsed value. If both those steps are successful, both values are
+||| returned as a pair.
+|||
```
Only the second value is returned here as opposed to `andbind`.

andrevidela replied:
Fixed, can you tell me if this rewording is acceptable?

gallais replied:
:+1:

andrevidela commented:
@gallais I replaced the implementation by the suggested one.

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
 roptand p q = map snd (optand p q)
 
+||| Attempts to parse using two different parsers.
+|||
```
May be worth to add "and remember where the success came from"

andrevidela replied:
I tried that but it made the first sentence too long. I've modified both the description sentence and the paragraph below to accommodate for your feedback, can you tell me if this is acceptable?

gallais replied:
:+1:

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
+|||
+||| Given an _opening_ parser, a _closing_ parser and a _middle_ parser this
+||| parser will succeed as long as _opening_, _middle_ and _cloding_ parser
+||| all succeed when executed in that order. Only the _middle_ parser will
```
Typo: cloding -> closing

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
+||| Given two parser, take the first one that succeeds.
+|||
+||| If the first parser fails, the second one will be attempted, if the second
+||| one fails the whole parser fails. This is anologous to an `or` operation.
```
typo: anologous -> analogous

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
+||| Run a monadic computation and a parser in succession.
+|||
+||| Given a monad `M`, a monadic computation `M a` and a parser `Parser M a`
+||| this function will run the comutation `M a` first and then run the 
```
typo: comutation -> computation

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
+||| Like `mand` but discards the second value.
+|||
+||| Given a monad `M`, a monadic computation `M a` and a parser `Parser M a`
+||| this function will run the comutation `M a` first and then run the 
```
typo: comutation -> computation

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
+||| Like `mand` but discards the first value.
+|||
+||| Given a monad `M`, a monadic computation `M a` and a parser `Parser M a`
+||| this function will run the comutation `M a` first and then run the 
```
typo: comutation -> computation

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
 
+||| Given a token, return a parser that parses this token.
+|||
+||| Given a token `p`, return a parser that will only succeed if it encounter
```
encounter -> encounters

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
+||| Given a list of token, fail if the input starts with one of them.
+|||
+||| Given a list of rejected tokens, return a parser that will succeed only
+||| if it does not encounter one of the tokens in the input. Conversly, it
```
typo: Conversly -> Conversely

andrevidela replied:
fixed

clayrat commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
+||| Parse anything that might be between two other parsers.
+|||
+||| Given an _opening_ parser, a _closing_ parser and a _middle_ parser this
+||| parser will succeed as long as _opening_, _middle_ and _cloding_ parser
```
cloding->closing

andrevidela replied:
fixed

andrevidela commented:
Is it ready to be merged?

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
                                         choiceMap pure (Success.guardM f s)
 
+||| Constrains a parsed to succeed only when a predicate holds.
+|||
```
Typo parsed -> parser

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
+||| Processes a token into a `Maybe` value.
+|||
+||| Given a function that map a parsed token into a Maybe a, this function
+||| returns a parser that fails when the token is mapped to `Nothing` and
```
Typo: function that map -> function that maps

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
 alt p q = MkParser $ \mlen, ts => runParser p mlen ts <|> runParser q mlen ts
 
+||| Given a list of parsers, use take the first one that succeeds, in order.
+|||
```
Typo: use take -> take

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -67,6 +129,14 @@ andoptbind p q = MkParser $ \mlen, ts =>
                     let combine = Success.map (Functor.map Just) . (Success.and sa)
                     (Functor.map combine (runParser (call (q (Value sa)) salen) lteRefl (Leftovers sa))) <|> pure (Success.map (flip MkPair Nothing) sa)
 
+||| Parse a value and processes it into another parser.
```
Typo: Parse -> Parses

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
 alts = foldr alt fail
 
+||| Parse a value and processes it into another parser.
+|||
```
Typo: Parse -> Parses

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
                    adjust (runParser (call (q (Value sa)) salen) lteRefl (Leftovers sa))
 
+||| Parse a values and processes it in the context of a monad.
+|||
```
Typo: Parse -> Parses

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
 ands ps = NEList.foldr1 (\ p, ps => map (uncurry (<+>)) (and p ps)) (Functor.map (map singleton) ps)
 
+||| Run a parser and a monadic computation in succession.
+|||
```
Typo: Run -> Runs

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
 optand p q = alt (and (map Just p) q) (map (MkPair Nothing) q)
 
+||| Parse a value and processes it into another parser.
+|||
```
Typo: Parse -> Parses

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
 anyOf ts = alts (map (\t => exact t) ts)
 
+||| Parse anything between two other parsers.
+|||
```
Typo: Parse -> Parses

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -76,128 +146,341 @@ andbind p q = MkParser $ \mlen, ts =>
 between open close p = land (rand open p) close
 
+||| Parse anything that might be between two other parsers.
+|||
```
Typo: Parse -> Parses

gallais commented on src/TParsec/Combinators.idr:
```diff
@@ -260,6 +543,14 @@ chainr1 : (Alternative mn, Monad mn) =>
 chainr1 p op = hchainr p op p
 
+||| Given a parser, parse a non-empty list of it.
+|||
```
Typo: parse -> parses

andrevidela commented:
So I've looked into it and changed most of the phrasing to use the present tense rather than the imperative tense ("runs" instead of "run"). I've also changed the phrasing "this parses X"  for a contaminator rather than use the verbose (but more correct) formulation of "this function returns a parse that parses X"
